security/l1-contracts-dev-flow-vision.md

```diff
@@ -0,0 +1,159 @@
+anyway for completeness.
+
+## Testing
+
+In the above section we described how standard unit tests would be written during development. Another
+aspect is how deploy and upgrade scripts are written and tested during development, as this is a key
+change.
+
+Originally, `Deploy.s.sol` was the script that run as part of test `setUp`, so we can run tests
+against the state resulting from the deploy script. In practice, that `Deploy.s.sol` script was
+never used for real deploys. So `Deploy.s.sol` has been refactored to instead use OPCM to deploy
+a chain's L1 contracts and use that fresh deploy as the base for tests. This means that all tests
+will be known to pass against the state resulting after a new chain deploy. (In the future, we may
+delete the full `Deploy.s.sol` file and replace it with OPCM code directly, as embedding OPCM into
+this legacy deploy script was only done as it's a simpler transition).
+
+OPCM contains a `deploy()` method that takes some arguments to deploy a chain. Some changes to
+source code will require no changes to this method—for example, if you only change things like
+code comments or function internals. Other changes will require changes to this method, like
+adding new contracts or modifying the `initialize` signature. For the latter case, because OPCM
+is what's used to scaffold chains for testing, you will be forced to modify the `deploy` method to
+account for your changes in order to test them. This is great, as it's now trivial to keep deploy
+scripts up to date, and we essentially get this for free.
+
+OPCM will also contain an `upgrade` method as described in the [OPCM Upgrades design doc](l1-upgrades.md).
+Some basic unit tests will test this method. Additionally, a flag can be set to have tests
+run against the state resulting from running `OPCM.upgrade` against an existing chain (instead of the
+default where tests are run against the state resulting from a fresh deploy via `OPCM.deploy()`). CI will be
+responsible for running the tests against the `upgrade` method for all mainnet and sepolia chains
+that the Security Council + Optimism Foundation (the 2/2) hold keys for. This enforces that
+developers are also writing the upgrade logic at the same time, to provide guarantees that the full
+test suite passes for new deploys and upgrades for all existing chains.
+
+## Releases and Contract Deployments
+
+After development, we are now ready to cut a release candidate and deploy the new implementation
+contracts to a chain, such as devnet, sepolia, or mainnet. All contract deployments must be done
+against a tag of the form `op-contracts/vX.Y.Z[-rc.n]`. To deploy the contract, create a `bootstrap`
+command in op-deployer, and run that new command to deploy the new contracts.
+
+An alternate approach that we might support that avoids the need for bootstrap commands is using
+CREATE2 in the `DeploySuperchain.s.sol` and `DeployImplementations.s.sol` scripts. This way,
```

Comment:
I like this idea - there'd be just one bootstrap command to deploy the implementations in this case. We should take care to minimize the amount of configuration required to deploy the implementations so that deploying individual contracts is easy.